src/gurt/hash.c

```diff
@@ -537,16 +539,23 @@ d_hash_rec_insert_anonym(struct d_hash_table *htable, d_list_t *link,
 
 	/* has no key, hash table should have provided key generator */
 	ch_key_init(htable, link, arg);
-
 	idx = ch_rec_hash(htable, link);
 	bucket = &htable->ht_buckets[idx];
+
+	if (need_lock && !need_keyinit_lock)
+		ch_bucket_lock(htable, idx, false);
```

Comment:
I think we can have per-bucket cookie for key generation (high-bits are bucket ID, generated ID is still unique), so we don't have to lock all buckets, instead just randomly pick one of the bucket and lock it for key generation.

Reply:
> The dfuse test took 39 minutes with this PR, the last few master builds have been 32, 32, 32, 30, 32, 30 so I think this needs more work, a 20% reduction in dfuse performance is not acceptable.
> 
> https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-8704/26/testReport/FTEST_dfuse/DaosBuild/
> 
> https://build.hpdd.intel.com/job/daos-stack/job/daos/job/master/7267/testReport/FTEST_dfuse/DaosBuild/Test_Hardware___Functional_Hardware_Small___4___dfuse_daos_build_py_DaosBuild_test_dfuse_daos_build_nocache_run_container_dfuse_hosts_pool_server_config_servers_7942/ https://build.hpdd.intel.com/job/daos-stack/job/daos/job/master/7267/testReport/FTEST_dfuse/DaosBuild/Test_Hardware___Functional_Hardware_Small___4___dfuse_daos_build_py_DaosBuild_test_dfuse_daos_build_nocache_run_container_dfuse_hosts_pool_server_config_servers_7942/ https://build.hpdd.intel.com/job/daos-stack/job/daos/job/master/7266/testReport/FTEST_dfuse/DaosBuild/Test_Hardware___Functional_Hardware_Small___4___dfuse_daos_build_py_DaosBuild_test_dfuse_daos_build_nocache_run_container_dfuse_hosts_pool_server_config_servers_7c3a/ https://build.hpdd.intel.com/job/daos-stack/job/daos/job/master/7265/testReport/FTEST_dfuse/DaosBuild/Test_Hardware___Functional_Hardware_Small___4___dfuse_daos_build_py_DaosBuild_test_dfuse_daos_build_nocache_run_container_dfuse_hosts_pool_server_config_servers_7c3a/ https://build.hpdd.intel.com/job/daos-stack/job/daos/job/master/7259/testReport/FTEST_dfuse/DaosBuild/Test_Hardware___Functional_Hardware_Small___4___dfuse_daos_build_py_DaosBuild_test_dfuse_daos_build_nocache_run_container_dfuse_hosts_pool_server_config_servers_7c3a/ https://build.hpdd.intel.com/job/daos-stack/job/daos/job/master/7255/testReport/FTEST_dfuse/DaosBuild/Test_Hardware___Functional_Hardware_Small___4___dfuse_daos_build_py_DaosBuild_test_dfuse_daos_build_nocache_run_container_dfuse_hosts_pool_server_config_servers_7c3a/

Trying to understand why it caused performance drop, i am wondering whether limiting number of RW locks are needed
push a debug test PR here: https://github.com/daos-stack/daos/pull/10003

Reply:
Yup, that is a good idea.

Now we changed cookie to atomic, which means we don't need lock all buckets, i guess why these locks are kept there is because whether it is needed depends on implementation of hop_key_init().

For now, these locks won't be held any more.

Reply:
there is a feature bit D_HASH_FT_NO_KEYINIT_LOCK added, that will avoid take every bucket's lock.